design_docs/shell.md

```diff
@@ -0,0 +1,101 @@
+# Protostar Shell MVP
+
+## Table of Contents
+
+<!-- TOC -->
+* [Protostar Shell MVP](#protostar-shell-mvp)
+  * [Table of Contents](#table-of-contents)
+  * [Terminology](#terminology)
+  * [Context](#context)
+  * [Goal](#goal)
+  * [Architecture](#architecture)
+    * [Communication](#communication)
+    * [Shell](#shell)
+    * [CLI](#cli)
+    * [Configuration Parser](#configuration-parser)
+    * [Scarb Metadata Parser](#scarb-metadata-parser)
+    * [Backend Runner](#backend-runner)
+    * [Changes to Backends](#changes-to-backends)
+  * [Scope Limitations](#scope-limitations)
+<!-- TOC -->
+
+## Terminology
+
+- Backend - binary providing a part of protostar functionality
+- Message - arguments provided to the Backend
+
+## Context
+
+As defined in [Protostar modular architecture](protostar-architecture.md), to accelerate
+the development of Protostar and simplify integration with [Scarb](https://github.com/software-mansion/scarb)
+the whole application will be partitioned into separate smaller binaries.
+A layer above these binaries is required to facilitate both running them and, in the future,
+simplify communication between Scarb and Protostar.
+
+## Goal
+
+Create a separate Rust application that will provide these features
+
+- Parsing the inputted commands
+- Running Protostar binaries
+- Resolve the environment details from `protostar.toml` (`Scarb.toml` + `scarb metadata` in the future)
+- Providing the binaries with necessary arguments
+
+## Architecture
+
+### Communication
+
+Protostar Shell will a use a defined message format (like JSON) to provide commands
+and environment details to the Backends.
+
+Details of the format should be decided in revision of this document or in the separate document.
+
+### Shell
+
+The core part of the Protostar Shell will be a base shell that will be able to execute
+any binary and parse any commands provided the required configuration.
+
+### CLI
+
+Translate CLI arguments to the Backend message format based on the configuration provided
+to the [Shell](#shell).
+
+Initially CLI will be written by-hand to only support `protostar test` command. Other commands
+will be simply passed directly to Protostar.
+
+_In the future, better abstraction over these commands can be created. However, eventually all
+CLI handling and calling corresponding backends will be done on Protostar Shell side -
+Backends will only accept JSON commands. That may result in CLI having to be written by-hand
+for most (or all) commands._
+
+### Configuration Parser
+
+Parse configuration file (`protostar.toml` and, in the future, `Scarb.toml`) and resolve
+details required for running Backends such as network configuration, contracts' dependencies, etc.
+
+### Scarb Metadata Parser
+
+Parse the output of `scarb metadata` command and resolve details required for running Backends.
+
+### Backend Runner
+
+Run Backends as separate subprocesses and provide them with necessary Messages.
```

Comment:
I think we might wanna use `execvp` mechanism here, that I also plan to use in Scarb, but we haven't yet got to implementing it properly: https://github.com/software-mansion/scarb/blob/5b6ca0d1e8b24cfe0bbff52506de9a75f1ec1bde/scarb/src/process.rs#L15-L31

Reply:
Not sure if we want to commit to specific mechanism in this document: It's more of a implementation detail. I can change the wording of `separate subprocesses` so it doesn't suggest anything specific if you think it may be useful.